src/ESA.cpp

```diff
@@ -0,0 +1,232 @@
+    double dfxc_t_rs = mixed_derivative(theta, rs, dx);
+    
+    const int nx = wvg.size();
+    slfc.resize(nx);
+    // Loop over the wave vector grid size
+    for (int i=0; i<nx; ++i) {
+                
+        double AF = 1.0/2.0 * (1 + tanh(Eta * (wvg[i] - xm)));
+
+        double GCSR = -(M_PI/12) * lambda * rs * pow(wvg[i], 2.0) * (4 * pow(theta, 2.0) * dfxc_t2 + pow(rs, 2.0) * dfxc_rs2 + 4 * theta * rs * dfxc_t_rs - 2 * theta * dfxc_t - 2 * rs * dfxc_rs);
+                
+        double Gnnfit = (1 + a * wvg[i] + b * pow(wvg[i], 1.0/2.0))/(1 + c * wvg[i] + d * pow(wvg[i], 1.25) + GCSR);
+        
+        double Gesa = GCSR * Gnnfit * (1 - AF) + (1 - g00) * AF;
+        
+        slfc[i] = Gesa;
+    }
+    // Assign the old slfc values to the slfcOld vector
+    slfcOld = slfc;
+
+}
+
+// QMC free energy function constants
+void fi(double theta, double &fa, double &fb, double &fd, double &fe, double &fc) {
+    // Constant for unit conversion
+    const double lambda = pow(4.0/(9.0*M_PI), 1.0/3.0);
+
+    const double omega = 1;
+    const double fb1 = 0.3436902;
+    const double fb2 = 7.82159531356;
+    const double fb3 = 0.300483986662;
+    const double fb4 = 15.8443467125;
+    const double fb5 = fb3*pow(3.0/2.0, 1.0/2.0)*omega/lambda;
+    const double fc1 = 0.8759442;
+    const double fc2 = -0.230130843551;
+    const double fd1 = 0.72700876;
+    const double fd2 = 2.38264734144;
+    const double fd3 = 0.30221237251;
+    const double fd4 = 4.39347718395;
+    const double fd5 = 0.729951339845;
+    const double fe1 = 0.25388214;
+    const double fe2 = 0.815795138599;
+    const double fe3 = 0.0646844410481;
+    const double fe4 = 15.0984620477;
+    const double fe5 = 0.230761357474;
+
+    fa = 0.610887 * tanh(pow(theta, -1)) * (0.75 + 3.04363 * pow(theta, 2) - 0.09227 * pow(theta, 3) + 1.7035 * pow(theta, 4)) / (1 + 8.31051 * pow(theta, 2) + 5.1105 * pow(theta, 4));
+    fb = tanh(1 / sqrt(theta)) * (fb1 + fb2 * pow(theta, 2) + fb3 * pow(theta, 4)) / (1 + fb4 * pow(theta, 2) + fb5 * pow(theta, 4));
+    fd = tanh(1 / sqrt(theta)) * (fd1 + fd2 * pow(theta, 2) + fd3 * pow(theta, 4)) / (1 + fd4 * pow(theta, 2) + fd5 * pow(theta, 4));
+    fe = tanh(1 / theta) * (fe1 + fe2 * pow(theta, 2) + fe3 * pow(theta, 4)) / (1 + fe4 * pow(theta, 2) + fe5 * pow(theta, 4));
+    fc = (fc1 + fc2 * exp(-(1 / theta))) * fe;
+    }
+
+// QMC free energy fitting function
+double fxc(double theta, double rs) {
+    const double omega = 1;
+    double fa, fb, fd, fe, fc;
+    fi(theta, fa, fb, fd, fe, fc);
+    return -(1 / rs) * (omega * fa + fb * sqrt(rs) + fc * rs) / (1 + fd * sqrt(rs) + fe * rs);
+}
+
+// Computation of the derivatives of the free energy functional
+double ESA::derivative_wrt_rs(double theta, double rs, double dx) const{
+    return (fxc(theta, rs + dx) - fxc(theta, rs - dx)) / (2 * dx);
+}
+
+double ESA::derivative_wrt_theta(double theta, double rs, double dx) const{
```

Comment:
Access theta and rs from the input structure, make dx a data member of the ESA class so you don't have to pass as an input parameter. Since this method is a one-liner that is called only once in computeSlfc, you can also just remove the methods and copy the lines directly in computeSlfc